### PR TITLE
`tide`: preserve branch merge type ordering 

### PR DESCRIPTION
**Pull request by danilo-gemoli**
Due to #29039 branch merge types get sorted alphabetically before searching for a match. 
This is somehow unexpected from a user standpoint who wants to apply the following configuration:
```yaml
tide:
  merge_method:
    kubernetes:
      test-infra:
        master: merge
        '.+': squash
```
In the previous scenario I want `master` to be matched before `.+`, but because of sorting I get the opposite, that is the `.+` regex swallowing essentially everything.
This PR attempts to preserve the branches' ordering as it was specified by the user in the `.yaml` configuration file.

I have assessed several libraries as [gitlab.com/c0b/go-ordered-json](https://pkg.go.dev/gitlab.com/c0b/go-ordered-json) but I then came to the conclusion that a simple regex offers these pros:
- it does the job
- less code
- no need to import another library

/cc @SergeyKanzhelev @petr-muller @jmguzik

### CODE REVIEW

**Comment by petr-muller** [comment]
Sorry for not being able to check this one earlier. Thanks for taking a stab at this, I agree the sorting had some unexpected consequences. I am not entirely convinced that we need *this* fix though: enhancing yaml contract for respect ordering adds complexity, and don't think we (nor yaml) promised the ordering matters. The snippet you posted is IMO close to ambiguous config.

Personally I'd be more inclined to put in a preference for exact matches: sort, pass to search for exact match, pass again to find a regex match, or possibly not fix this at all (possibly adding a sentence to the godoc that if multiple items cover a repo then arbitrary one would be selected) . I don't hold this opinion too strongly.

**Comment by danilo-gemoli** [comment]
> enhancing yaml contract for respect ordering adds complexity

You are right, it does but at the same time the following proposed solution:

> Personally I'd be more inclined to put in a preference for exact matches: sort, pass to search for exact match, pass again to find a regex match, or possibly not fix this at all (possibly adding a sentence to the godoc that if multiple items cover a repo then arbitrary one would be selected)

adds complexity on the user end who simply wants to write a couple of entries down and expects the first (starting from the top) to that matches to be chosen.
I am open to other alternatives as shifting from a dictionary to an array like so:
```yaml
tide:
  merge_method:
    kubernetes:
      test-infra:
      - master: merge
      - '.+': squash
```
I've yet to check the `yaml` specification but I expect it to states  somewhere that an array is an **ordered** list of items and therefore such a property has to be preserved after the deserialization step.

In particular, this concerns me the most:
> if multiple items cover a repo then arbitrary one would be selected

**Comment by petr-muller** [comment]
> expects the first (starting from the top) to that matches to be chosen.

Not every expectation is well-founded, and not every expectation is an actual feature we want to implement and maintain.  I do not think that every advanced use case must be supported in the most convenient way for the user, and we do not need to accommodate every advanced use case.

**Comment by danilo-gemoli** [comment]
In this case it looks fine to me to leave this feature out and keep working on what is left to have the whole feature done.
Feel free to freeze/close/or-whatever-is-more-appropriate this PR. We could always come back here when and if we will need it.

**Comment by petr-muller** [comment]
Let's keep it open for a bit, maybe it attracts more opinions. I'm not the sole approver.

**Comment by SergeyKanzhelev** [comment]
key order is not guarantee seems like: https://yaml.org/spec/1.2-old/spec.html#id2765608:~:text=3.4.%C2%A0Serialization%20Model-,3.2.2.1.%C2%A0Keys%20Order,-In%20the%20representation I think many parsers do use it though. If ordering is needed, maybe this can be converted to a list?

I have no opinion beyond this note on the right way to implement this.

**Comment by droslean** [comment]
From my point of view, I am not convinced that we should address this. Also, I do not favor increasing the complexity of Tide's configuration.

File: prow/config/tide.go [hunk 1]
```diff
@@ -90,6 +104,17 @@ func (trmt *TideRepoMergeType) UnmarshalJSON(b []byte) error {
...
 	}
 	var branches map[string]TideBranchMergeType
+
+	branchMatches := countTideBranchesMergeMethodRegexp.FindAllStringSubmatch(string(b), -1)
+	trmt.BranchesOrder = make([]string, 0, len(branchMatches))
+	for _, branchSubmatch := range branchMatches {
+		if len(branchSubmatch) == 3 {
+			trmt.BranchesOrder = append(trmt.BranchesOrder, branchSubmatch[2])
+		} else {
+			return errors.New("TideRepoMergeType unmarshal: unable to determine branches order")
```
Comment (droslean) [inline prow/config/tide.go:114]: This has to be done when validating the tide config.

**Pull request closed without merging**